src/main/java/uk/gov/pay/directdebit/filters/LoggingMDCRequestFilter.java

```diff
@@ -0,0 +1,50 @@
+package uk.gov.pay.directdebit.filters;
+
+import org.slf4j.MDC;
+import uk.gov.pay.directdebit.gatewayaccounts.dao.GatewayAccountDao;
+
+import javax.inject.Inject;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import java.io.IOException;
+import java.util.Optional;
+
+import static uk.gov.pay.logging.LoggingKeys.GATEWAY_ACCOUNT_ID;
+import static uk.gov.pay.logging.LoggingKeys.GATEWAY_ACCOUNT_TYPE;
+import static uk.gov.pay.logging.LoggingKeys.MANDATE_EXTERNAL_ID;
+import static uk.gov.pay.logging.LoggingKeys.PAYMENT_EXTERNAL_ID;
+import static uk.gov.pay.logging.LoggingKeys.PROVIDER;
+
+public class LoggingMDCRequestFilter implements ContainerRequestFilter {
+
+    private final GatewayAccountDao gatewayAccountDao;
+
+    @Inject
+    public LoggingMDCRequestFilter(GatewayAccountDao gatewayAccountDao) {
+        this.gatewayAccountDao = gatewayAccountDao;
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
+        getPathParameterFromRequest("accountId", requestContext)
+                .flatMap(gatewayAccountDao::findByExternalId)
+                .ifPresent(gatewayAccount -> {
+                    MDC.put(GATEWAY_ACCOUNT_ID, gatewayAccount.getExternalId());
+                    MDC.put(PROVIDER, gatewayAccount.getPaymentProvider().name());
+                    MDC.put(GATEWAY_ACCOUNT_TYPE, gatewayAccount.getType().name());
+                });
+        
+        getPathParameterFromRequest("mandateExternalId", requestContext)
+                .ifPresent(mandateExternalId -> MDC.put(MANDATE_EXTERNAL_ID, mandateExternalId));
+
+        getPathParameterFromRequest("paymentExternalId", requestContext)
+                .ifPresent(mandateExternalId -> MDC.put(PAYMENT_EXTERNAL_ID, mandateExternalId));
+        
+        getPathParameterFromRequest("token", requestContext)
```

Comment:
I decided not to add 'token' as a key in pay-java-commons as it's not common across projects like the charge id or gateway account id is.

Reply:
we already have `SECURE_TOKEN` in pay-java-commons which you can use

Reply:
oops this comment should have been on line 45. fixed.